server/src/main/java/org/elasticsearch/common/lucene/uid/PerThreadIDVersionAndSeqNoLookup.java

```diff
@@ -137,19 +137,39 @@ private int getDocID(BytesRef id, Bits liveDocs) throws IOException {
     }
 
     /** Return null if id is not found. */
-    DocIdAndSeqNo lookupSeqNo(BytesRef id, LeafReaderContext context) throws IOException {
+    DocIdAndSeqNo lookupSeqNo(BytesRef id, LeafReaderContext context, long minSeqNoForDeletes, boolean includeDeletes) throws IOException {
         assert context.reader().getCoreCacheHelper().getKey().equals(readerKey) :
             "context's reader is not the same as the reader class was initialized on.";
-        int docID = getDocID(id, context.reader().getLiveDocs());
-        if (docID != DocIdSetIterator.NO_MORE_DOCS) {
-            NumericDocValues seqNos = context.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
-            long seqNo;
-            if (seqNos != null && seqNos.advanceExact(docID)) {
-                seqNo = seqNos.longValue();
-            } else {
-                seqNo =  SequenceNumbers.UNASSIGNED_SEQ_NO;
+        // termsEnum can possibly be null here if this leaf contains only no-ops.
+        if (termsEnum != null && termsEnum.seekExact(id)) {
+            docsEnum = termsEnum.postings(docsEnum, 0);
+            final Bits liveDocs = context.reader().getLiveDocs();
+            // there may be more than one matching docID, in the case of nested docs, so we want the last one( this isn't strictly
+            // required as we are no longer use primary_term as a tier breaker).
+            DocIdAndSeqNo result = null;
+            int docID = docsEnum.nextDoc();
+            if (docID != DocIdSetIterator.NO_MORE_DOCS) {
+                final NumericDocValues seqNoDV = context.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
+                for (; docID != DocIdSetIterator.NO_MORE_DOCS; docID = docsEnum.nextDoc()) {
+                    if (liveDocs == null || liveDocs.get(docID)) {
+                        final long seqNo;
+                        if (seqNoDV != null && seqNoDV.advanceExact(docID)) {
+                            seqNo = seqNoDV.longValue();
+                        } else {
+                            seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+                        }
+                        result = new DocIdAndSeqNo(docID, seqNo, context);
```

Comment:
is it possible that we have a live doc first with doc Id 1 and then a deleted doc with docId 2 that is > `minSeqNoForDeletes` will we use the delete instead of the live doc and is that correct? or should we prefer the live doc all the time? I am thinking of a situation where we have the deleted doc after the updated due to a merge?

Reply:
that' shouldn't be possible. A live do is always the one with the highest seq# (if exists).

Reply:
I think you are missing that the index is not ordered by seqID. I think we need to return the result immediately if found for a live doc?